cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/RawMetricValues.java

```diff
@@ -0,0 +1,351 @@
+    _imputations = new BitSet(numWindows);
+    _validity = new BitSet(numWindows);
+    _minSamplesPerSnapshotWindow = minSamplesPerSnapshotWindow;
+    _oldestWindowIndex = Long.MAX_VALUE;
+  }
+
+  public synchronized void addSample(MetricSample<?, ?> sample, long windowIndex, MetricDef metricDef) {
+    // This sample is being added during window rolling.
+    if (windowIndex < _oldestWindowIndex) {
+      return;
+    } else if (windowIndex > _oldestWindowIndex + _counts.length) {
+      throw new IllegalArgumentException("Cannot add sample to window index " + windowIndex + ", which is larger "
+                                             + "than the current window index " + currentWindowIndex());
+    }
+    int idx = (int) (windowIndex % _counts.length);
+    for (Map.Entry<Integer, Double> entry : sample.allMetrics().entrySet()) {
+      _values.computeIfAbsent(entry.getKey(), k -> new float[_counts.length]);
+      updateValue(entry.getValue(), metricDef.metricInfo(entry.getKey()), idx);
+    }
+    _counts[idx]++;
+    updateValidityAndImputation(idx);
+    if (_counts[idx] >= _minSamplesPerSnapshotWindow) {
+      _imputations.clear(idx);
+      if (idx != firstIdx()) {
+        int prevIdx = prevIdx(idx);
+        if (_counts[prevIdx] < halfMinRequiredSamples()) {
+          updateAvgAdjacent(prevIdx);
+        }
+      }
+      if (idx != lastIdx()) {
+        int nextIdx = nextIdx(idx);
+        if (_counts[nextIdx] < halfMinRequiredSamples()) {
+          updateAvgAdjacent(nextIdx);
+        }
+      }
+    }
+    LOG.trace("Added metric sample {} to window index {}, actual index is {}, current count : {}",
+              sample, windowIndex, idx, _counts[idx]);
+  }
+
+  public synchronized void updateOldestWindowIndex(long newOldestWindowIndex) {
+    long prevLastWindowIndex = currentWindowIndex() - 1;
+    _oldestWindowIndex = newOldestWindowIndex;
+    // Advancing the oldest window index will make the previous current window index become available to its
+    // neighbour index (i.e. the previous last index) for AVG_ADJACENT imputation. We don't need to update the
+    // previous current window index because it would be up to date during the addSample call.
+    if (prevLastWindowIndex >= _oldestWindowIndex) {
+      updateValidityAndImputation(handleWrapping(prevLastWindowIndex));
+    }
+  }
+
+  public synchronized boolean isValid(float maxAllowedImputation) {
+    int currentIdx = handleWrapping(currentWindowIndex());
+    int numValidIndexesAdjustment = _validity.get(handleWrapping(currentIdx)) ? 1 : 0;
+    boolean allIndexesValid = _validity.cardinality() - numValidIndexesAdjustment == _counts.length - 1;
+
+    boolean tooManyImputations = numImputations() > maxAllowedImputation;
+
+    return allIndexesValid && !tooManyImputations;
+  }
+
+  public synchronized int numImputations() {
+    int currentIdx = handleWrapping(currentWindowIndex());
+    int numImputationAdjustment = _imputations.get(currentIdx) ? 1 : 0;
```

Comment:
The current window is still being updated. e.g. if the current window has just been rolled out, it will always be either invalid or flawed, so we should exclude it for now. We may reconsider after we decide to expose it to the users when there is only the current window available.